src/backend/distributed/utils/replication_origin_session_utils.c

```diff
@@ -0,0 +1,283 @@
+ *   Functions for managing replication origin session.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "distributed/replication_origin_session_utils.h"
+#include "distributed/remote_commands.h"
+#include "distributed/metadata_cache.h"
+#include "utils/builtins.h"
+#include "miscadmin.h"
+
+static bool IsRemoteReplicationOriginSessionSetup(MultiConnection *connection);
+
+static bool ExecuteRemoteCommandAndCheckResult(MultiConnection *connection, char *command,
+											   char *expected);
+
+static void SetupMemoryContextResetReplicationOriginHandler(void);
+
+static void SetupReplicationOriginSessionHelper(bool isContexResetSetupNeeded);
+
+static inline bool IsLocalReplicationOriginSessionActive(void);
+
+PG_FUNCTION_INFO_V1(citus_internal_start_replication_origin_tracking);
+PG_FUNCTION_INFO_V1(citus_internal_stop_replication_origin_tracking);
+PG_FUNCTION_INFO_V1(citus_internal_is_replication_origin_tracking_active);
+
+/*
+ * This variable is used to remember the replication origin id of the current session
+ * before resetting it to DoNotReplicateId in SetupReplicationOriginLocalSession.
+ */
+static RepOriginId OriginalOriginId = InvalidRepOriginId;
+
+/*
+ * Setting that controls whether replication origin tracking is enabled
+ */
+bool enable_change_data_capture = false;
+
+
+/* citus_internal_start_replication_origin_tracking starts a new replication origin session
+ * in the local node. This function is used to avoid publishing the WAL records to the
+ * replication slot by setting replication origin to DoNotReplicateId in WAL records.
+ * It remembers the previous replication origin for the current session which will be
+ * used to reset the replication origin to the previous value when the session ends.
+ */
+Datum
+citus_internal_start_replication_origin_tracking(PG_FUNCTION_ARGS)
+{
+	if (!enable_change_data_capture)
+	{
+		PG_RETURN_VOID();
+	}
+	SetupReplicationOriginSessionHelper(false);
+	PG_RETURN_VOID();
+}
+
+
+/* citus_internal_stop_replication_origin_tracking ends the current replication origin session
+ * in the local node. This function is used to reset the replication origin to the
+ * earlier value of replication origin.
+ */
+Datum
+citus_internal_stop_replication_origin_tracking(PG_FUNCTION_ARGS)
+{
+	ResetReplicationOriginLocalSession();
+	PG_RETURN_VOID();
+}
+
+
+/* citus_internal_is_replication_origin_tracking_active checks if the current replication origin
+ * session is active in the local node.
+ */
+Datum
+citus_internal_is_replication_origin_tracking_active(PG_FUNCTION_ARGS)
+{
+	bool result = IsLocalReplicationOriginSessionActive();
+	PG_RETURN_BOOL(result);
+}
+
+
+/* IsLocalReplicationOriginSessionActive checks if the current replication origin
+ * session is active in the local node.
+ */
+inline bool
+IsLocalReplicationOriginSessionActive(void)
+{
+	return (replorigin_session_origin != InvalidRepOriginId);
```

Comment:
Should not it be `replorigin_session_origin != DoNotReplicateId`?

Reply:
Generally the if replication origin is set, it should be filtered by this callback..
But my CDC changes set it to "DoNotReplicateId" 0xFFFF.
@marcocitus  please let me know your opinion on this.

Reply:
seems like that should be a separate function

I would expect citus_internal_is_replication_origin_tracking_active to check for DoNotReplicateId

Reply:
Ok